[SLO] Group filters in the Overview embeddable

mgiota opened the pull request:
Fixes https://github.com/elastic/kibana/issues/179908
Fixes https://github.com/elastic/kibana/issues/179895



https://github.com/elastic/kibana/assets/2852703/2924e619-9080-4c80-b5c3-7066a100026c



## UI changes

### SLO Overview configuration
<img width="600" alt="Screenshot 2024-04-15 at 13 34 21" src="https://github.com/elastic/kibana/assets/2852703/3635c9e7-e5c1-454b-bb60-93b5d637083e">


### Rendered panel with Group of SLOs
<img width="1022" alt="Screenshot 2024-04-03 at 13 08 11" src="https://github.com/elastic/kibana/assets/2852703/377c5cd0-b6e0-4906-888b-93893a5f1208">

### Updated SLO list group by page
<img width="1316" alt="Screenshot 2024-03-28 at 23 56 12" src="https://github.com/elastic/kibana/assets/2852703/d6b306a0-77e8-44e5-b103-ce3537b0ceda">

## Responsiveness
This PR makes sure that the UI doesn't look broken in smaller screen resolutions and fixes following issues. If we want to display things differently in smaller screen resolutions, we should handle in a follow up PR (cc @maciejforcone)

### Before 

####  SLO list page in smaller screen resolution
<img width="500" alt="Screenshot 2024-04-03 at 10 14 31" src="https://github.com/elastic/kibana/assets/2852703/f2b11408-3356-4f4b-9c4b-ec085ff494fa">

####  Rendered panel with Group of SLOs in smaller screen resolution
<img width="500" alt="Screenshot 2024-04-03 at 10 13 18" src="https://github.com/elastic/kibana/assets/2852703/6e37ad61-5a0a-4027-8b90-4f8fad243637">

#### After

####  SLO list page in smaller screen resolution
<img width="500" alt="Screenshot 2024-04-03 at 11 04 35" src="https://github.com/elastic/kibana/assets/2852703/fa875aa8-bc54-4c6b-9159-4462e585a9cb">

####  Rendered panel with Group of SLOs in smaller screen resolution
I applied a `min-width` of 100px to the SLO name, so that it doesn't look broken. What is not visible in the screenshot is a horizontal scrollbar, so user can still see the part with Worst performing that is not visible. 
<img width="500" alt="Screenshot 2024-04-03 at 10 13 47" src="https://github.com/elastic/kibana/assets/2852703/d9566ca3-42af-4d85-a3f7-6374cc27981d">


## Acceptance criteria
- Overview configuration
  -  toggles between Single SLO or Grouped SLOs
  - Required field for the `Grouped SLOs` view is only the `Group by` option, the rest filters are optional
  - Users can optionally select the groups they want
  - Users can use the KQL bar to apply extra filters
- Overview configuration opens in a `Flyout` (modal was causing some issues with z-index and kqlbar suggestions being displayed below the overlay)
- Embeddable panel
 - The rendered panel shows the selected groups with summary data divided with a horizontal like and no border around it
 - Clicking on a group opens an accordion similar to the SLO list group by page
 - An Edit criteria link is visible only in the Grouped Embeddable panel
 - Clicking on the Edit criteria link opens the configuration modal, showing only the grouped tab and hiding the tabbed button group
 - The grouped embeddable should react to Dashboard's Refresh button
 - `Add to Dashboard` action should be hidden in Dashboard context

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `/oblt-deploy-serverless` : Deploy a serverless Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

elasticmachine commented:
Pinging @elastic/obs-ux-management-team (Team:obs-ux-management)

mgiota commented on x-pack/plugins/observability_solution/slo/common/locators/paths.ts:
```diff
@@ -25,8 +25,8 @@ export const paths = {
   sloDetails: (sloId: string, instanceId?: string) =>
     !!instanceId
-      ? `${SLOS_BASE_PATH}${SLOS_PATH}/${encodeURIComponent(sloId)}?instanceId=${encodeURIComponent(
+      ? `${SLOS_BASE_PATH}/${encodeURIComponent(sloId)}?instanceId=${encodeURIComponent(
```
This was adding an extra `/` in the SLO details path (`http://{KIBANA_URL}/app/slos//{SLO_ID}`) of the `Worst performing` link when trying to navigate from the dashboard app to the slo detail page. 

After this fix, worst performing link works fine in both pages (Dashboard app and SLO list page)!

https://github.com/elastic/kibana/assets/2852703/ee72b14a-fd56-4a87-82fa-a30091ca555d

mgiota commented:
@elasticmachine merge upstream

mgiota commented:
@elasticmachine merge upstream

simianhacker commented:
Cool feature... I'd like to be able to add the panel where I don't have to select any of the groupings ahead of time. For example, what if wanted to add a panel to group by status but all of my clusters are currently healthy? When something is degrading or violated, I'd like for that grouping to automatically show up. 

I think selecting the "Tags", "Statuses", and "SLI Type" should be optional.

simianhacker commented on x-pack/plugins/observability_solution/slo/public/embeddable/slo/overview/slo_embeddable.tsx:
```diff
@@ -75,30 +82,59 @@ export class SLOEmbeddable extends AbstractEmbeddable<SloEmbeddableInput, Embedd
     const queryClient = new QueryClient();
-
+    const { observabilityRuleTypeRegistry } = this.deps.observability;
+    // const observabilityRuleTypeRegistry = this.deps.observabilityRuleTypeRegistry;
```
Can we remove this comment?

mgiota replied:
@mgiota I removed it

mgiota commented:
> I think selecting the "Tags", "Statuses", and "SLI Type" should be optional.

@simianhacker Very good point! Do you mind if I handle it in a separate PR, since I need a few clarifications regarding following: 

- How will the user flow be like? 

I can remove the validation that I currently have in the second dropdown (the groupings) and have all options selected by default. What do you think? @maciejforcone I would like your input here.

Currently I display only "healthy" statuses if I have for example only healthy SLOs, and you suggest to display all statuses which comes to my second question.
<img width="500" alt="319157520-28482001-3caa-4912-9f4a-b28554966b76" src="https://github.com/elastic/kibana/assets/2852703/a4a0c91f-4b87-4013-a5cd-455488cb1ddf">

- Is it OK if I apply the same change to the SLO list page as well? Because there we have same thing happening. When we select group by status and we have only healthy SLOs, then only healthy group will appear. I would like to apply the change to one place and it will affect both SLO list page and SLO Overview embeddable.

mgiota commented:
Apart from what you suggest, I was thinking that it now makes sense to add an `Edit settings` action to the Overview Embeddable, where user can change their selected configuration. I will create a separate issue for this. For reference I am linking here the [implementation](https://github.com/elastic/kibana/pull/169910/commits/80995f25bbee45bfb9daf44834230db111496b98) of the Edit settings in the alerts embeddable  and the [PR](https://github.com/elastic/kibana/pull/178844) for decoupling the action from the Embeddable framework.

mgiota commented:
@elasticmachine merge upstream

mgiota commented:
As an enhancement, I could include the `selected group in the panel's title` as well. If user has a few embeddables in their dashboards grouped by different group by options, it would be more clear if each panel showed a more specific title for example `SLO overview grouped by Tags` or `SLO overview grouped by Status` accordingly. What do you think?

mgiota commented:
@elasticmachine merge upstream

mgiota commented:
@elasticmachine merge upstream

mgiota commented:
@elasticmachine merge upstream

mgiota commented:
After discussing with @maciejforcone we decided to completely remove the second dropdown (selection of groups). This way a new grouping will automatically show up as shown below. Here's how it looks like so far:

https://github.com/elastic/kibana/assets/2852703/3496ce1f-64b9-4152-9e17-8b57e0052c1b

What I am going to add as well is a `KQL search bar` along with `tags filtering` as we currently have in the Overview page. I will upload a Video once I have the implementation ready. Let me know if something doesn't feel right so far.

mgiota commented:
@elasticmachine merge upstream

mgiota commented:
Just wanted to share the progress so far.

I removed the second dropdown from the modal and I updated the panel's title to include the group by. 
<img width="1495" alt="Screenshot 2024-04-09 at 08 59 37" src="https://github.com/elastic/kibana/assets/2852703/8dd6c04b-2c39-469b-b4f8-adde3b08fb0c">

I am working now on adding the KQL bar, but I am facing at the moment a UI issue with the z-index of the suggestions box.

<img width="400" alt="Screenshot 2024-04-09 at 10 02 57" src="https://github.com/elastic/kibana/assets/2852703/5b0e9d7c-b69c-45e4-92ac-ed33ebde7519">

After I fix the UI issue, we need one more change to consider the feature complete. We need to add an `Edit settings` action in the panels context menu, so that users can open the modal and see their selection. Apart from that, I think we would need to add in the rendered panel an extra visual indication of user's selection. Easiest solution would be to have a `Settings` link at the top right of the panel that opens the Edit configuration modal.

jasonrhodes commented:
Great work on this, @mgiota -- just some very high level questions below, CC: @maciejforcone / @grabowskit 

[1] Are you all comfortable with losing the simpler filtering feature in this MVP, without the more powerful KQL filter option added? Will people with lots of SLOs find this valuable since it requires them to show ALL SLOs in this form? 

[2] Should we rename that tab from "Group of SLOs" to "Grouped SLOs" or "Multiple SLOs"?

mgiota commented:
@elasticmachine merge upstream

mgiota commented on x-pack/plugins/observability_solution/slo/public/embeddable/slo/overview/slo_configuration.scss:
```diff
@@ -0,0 +1,3 @@
+.euiOverlayMask {
+  z-index: unset !important
```
This fixes the issue with the kqlbar suggestions shown below the overlay mask of the modal. 

<img width="773" alt="320770685-5b0e9d7c-b69c-45e4-92ac-ed33ebde7519" src="https://github.com/elastic/kibana/assets/2852703/34e51b65-c8dc-43d5-8e8a-da1eb0661580">

Similar to this [one](https://github.com/elastic/eui/issues/5757). Suggested [solution](https://github.com/elastic/eui/issues/5757#issuecomment-1088265561)

mgiota replied:
@andreadelrio Thanks for looking into this one. You are right, I had it in my todo list to fix the issue with the custom filter. I just pushed the change. Let me look at the button group issue as well

mgiota replied:
I fixed the `button-group` padding issue as well. Let me know what you think!

<img width="400" alt="Screenshot 2024-04-11 at 21 59 10" src="https://github.com/elastic/kibana/assets/2852703/e3b86a80-bdbd-4efb-a3e0-f132839f3e46">

andreadelrio replied:
@mgiota Thanks for fixing that. This override is generating some weird consequences such as:

a) The top navigation is coming to the front of the overlay when it shouldn't
<img width="1077" alt="image" src="https://github.com/elastic/kibana/assets/4016496/81a013c4-2600-4c5e-a5c5-d16064e1c373">
<img width="1440" alt="image" src="https://github.com/elastic/kibana/assets/4016496/ea03da1f-9028-4100-a8e1-657e5b2421a6">

b) In serverless the flyout's header is covered by the navigation.
<img width="1442" alt="image" src="https://github.com/elastic/kibana/assets/4016496/0537d65f-7fa8-402f-8e51-695cf880c1cf">

You might want to consult with @elastic/eui-team if there's a better way to achieve what you are trying to achieve.

mgiota replied:
@andreadelrio Thanks a lot for looking into it thoroughly! I fixed the issue for both cases you described:

<img width="400" alt="Screenshot 2024-04-12 at 11 25 09" src="https://github.com/elastic/kibana/assets/2852703/00724c88-5562-498b-85f6-6b9dd190016a">


<img width="400" alt="Screenshot 2024-04-12 at 11 23 10" src="https://github.com/elastic/kibana/assets/2852703/6b0ab221-8170-4986-9620-e6bcc93a4083">

I definitely plan to reach out to @elastic/eui-team and possibly unified Search team, since I have more questions for them. In a nutshell, I had 2 options for the initial problem with the kql bar suggestions showing below the overlay mask:
- either increase the z-index of the `Searchbar`. Unfortunately using `css` directly in the unified `SearchBar` component is not possible. The whole page was breaking, telling me that SearchBar is not defined. So I left this solution, which I would ideally prefer to use.
- or decrease the z-index of the overlayMask to make suggestions bar visible. This is the approach I followed, now it works with the z-index I provide


Hopefully I can get your approval. It is a cool feature we would like to end up in the following release. Thanks for your support!

cee-chen replied:
⚠️ Please don't override the CSS of EUI global components without adding some sort of modifier that clearly indicates you only want to apply it to **this specific usage** - otherwise the change cascades down everywhere (and I believe to multiple plugins??) and causes a ton of side effects.

[Modals are meant to overlay on top of **everything**](https://eui.elastic.co/#/layout/modal/guidelines) including the global nav. They're essentially interrupt-driven dialogs. If you don't want this behavior, you should not be using an EuiModal - please consider using an [EuiFlyout](https://eui.elastic.co/#/layout/flyout) instead, which automatically sits below the global nav bar and leaves it accessible to keyboard **and** mouse users.

andreadelrio replied:
++ to Cee's comments. @mgiota I believe a flyout would be better suited here. This is the component of choice for most of our panels (with the exception of those that have a dedicated editor). You can also refer to this internal [document](https://docs.google.com/document/d/1vNcHT3YgX0bZNGLILS1LtlP6nW2z806ryU2ESgYRpAM/edit?usp=sharing) where we make this recommendation `Best practices for teams adding new panels to Dashboards`

>Creation: ideally users should be able to create the new panel by displaying a flyout in the same dashboard. If this is not possible, they should find at least a link to creation (see ML example below)`. cc @elastic/kibana-presentation for awareness.

mgiota replied:
cc @jasonrhodes This is the issue I mentioned to you with the z-index and the modal. I couldn't find a way to fix the issue of the kqlbar fixing positioned below the modal mask, without overriding the CSS of the global component. 

@andreadelrio and @cee-chen We were inspired by Machine learning embeddable regarding the use of Modal for our embeddable's configuration. But our use case is a bit more complex, since now we want to include a kql bar which causes the issue.

<img width="400" alt="Screenshot 2024-04-15 at 08 37 54" src="https://github.com/elastic/kibana/assets/2852703/1235fc08-c040-431f-945d-06930ae0cc5b">


As you described above the overlay is positioned on top of everything. I was trying to find a way to apply my change to a specific usage, but I couldn't. More specifically this is the solution I was hoping would work, but unfortunately this breaks the whole embeddable configuration. Modal doesn't even open. That's why I went with the solution of overriding the global css (making sure I don't break anything in the Dashboard app with manual testing on my side). I was expecting design team to comment on this and recommend me a better way to do this. Changing to Flyout sounds good, but it sounds like a bigger change that I was hoping for.


<img width="400" alt="Screenshot 2024-04-15 at 08 52 53" src="https://github.com/elastic/kibana/assets/2852703/2f728db9-3960-461b-ac4d-9f2d0980b055">


@jasonrhodes What options do we have here?
- remove the use of kqlbar so that we ship the grouping functionality to our users?
- replace modal with flyout?
- try to find a way to apply the change I want to the kqlbar?

The ideal solution is the 3rd one, I just couldn't find a way to do it. From my understanding this would involve unified search team and the option to support css on their component. @andreadelrio and @cee-chen Do you have something else to recommend on how to fix the z-Index issue?

mgiota replied:
@andreadelrio @cee-chen I pushed the final fix, where I replace `Modal` with `Flyout` according to your suggestions. 

https://github.com/elastic/kibana/assets/2852703/c2975bb8-616a-4e1c-9a3f-1b486e1f232c

Let me know if it looks good and if we can get final approval for this PR.

cee-chen replied:
It looks terrific to me! I have a minor spacing suggestion for the save/cancel button locations that I'll make in a separate code commit (the below is from our docs and is likely what most flyouts use across Kibana).

<img width="756" alt="" src="https://github.com/elastic/kibana/assets/549407/4a20aa62-b4bd-4600-ad81-56192dc556bf">

mgiota commented:
> @maciejforcone We talked about having Single SLO and Groups of SLOs in a single embeddable in the past. Now that I have to handle more scenarios, when I switch from one view to the other, I am just wondering if it feels right to mix these two in a single embeddable.


I fixed this issue by simply allowing to edit the criteria only in the grouped SLOs and hiding the tabbed buttons on the edit screen.

<img width="400" alt="Screenshot 2024-04-11 at 19 04 00" src="https://github.com/elastic/kibana/assets/2852703/0718301e-90f2-4a27-8075-cd45cc2f44ba">

mgiota commented:
> [1] Are you all comfortable with losing the simpler filtering feature in this MVP, without the more powerful KQL filter option added? Will people with lots of SLOs find this valuable since it requires them to show ALL SLOs in this form?

I included both groups filtering and KQL in this version! 

https://github.com/elastic/kibana/assets/2852703/bc161444-b6fd-4640-be5c-6375eb131441



> [2] Should we rename that tab from "Group of SLOs" to "Grouped SLOs" or "Multiple SLOs"?

I renamed to `Grouped SLOs`

mgiota commented:
@elasticmachine merge upstream

simianhacker commented on x-pack/plugins/observability_solution/slo/public/embeddable/slo/overview/group_view.tsx:
```diff
@@ -0,0 +1,84 @@
+  useEffect(() => {
+    const subs = reloadGroupSubject?.subscribe((input) => {
+      if (input) {
+        const ngroupBy = input?.groupFilters?.groupBy ?? groupBy;
+        setGroupBy(ngroupBy);
```
```suggestion
        const nGroupBy = input?.groupFilters?.groupBy ?? groupBy;
        setGroupBy(nGroupBy);
```

mgiota replied:
I updated it!

simianhacker reviewed (changes requested):
I added an embeddable to the dashboard with:

- Group by: Tags
- Tags: (none selected)
- Filter: `slo.instanceId: "blast-mail.co"` 

Which rendered as expected:

<img width="592" alt="image" src="https://github.com/elastic/kibana/assets/41702/22bf1e4f-33dd-4da3-a41a-27864adcfcee">

Then I edited the it to only select the `admin-console` expecting the group by to only return just the `admin-console` but it still rendered the same as above with the `frontend` tag as well.

I'm wondering if the expectation would be to only display the tags returned that match the list of tags you picked even though the SLOs returned also belong to the `frontend` tag OR render "as-is" in this implementation? @maciejforcone thoughts?

I filed an issue about how we should hide the pagination when there are less items than the number of items per page: https://github.com/elastic/kibana/issues/180642 which can be addressed in a follow up PR since it also affects the "SLO Overview" page.

BTW... there's one change I suggested for consistency sake.

mgiota commented:
> I'm wondering if the expectation would be to only display the tags returned that match the list of tags you picked even though the SLOs returned also belong to the `frontend` tag OR render "as-is" in this implementation? @maciejforcone thoughts?

@simianhacker Yep I have seen same behavior before in SLO overview page and in the beginning it looked odd. Let's see what @maciejforcone thinks about it. We could cover this in a follow up PR, since it is how it currently works on SLO Overview as well.

mgiota commented:
@elasticmachine merge upstream

mgiota commented:
@elasticmachine merge upstream

cee-chen commented on x-pack/plugins/observability_solution/slo/public/embeddable/slo/overview/slo_configuration.tsx:
```diff
@@ -7,81 +7,106 @@
-      </EuiModalBody>
-      <EuiModalFooter>
+      </EuiFlyoutBody>
+      <EuiFlyoutFooter>
```
Minor suggestion for spacing out the cancel and save buttons:

```suggestion
      <EuiFlyoutFooter>
        <EuiFlexGroup justifyContent="spaceBetween">
```

cee-chen commented:
LGTM from an EUI usage perspective, thank you so much for the refactor from EuiModal to EuiFlyout! Will defer to @andreadelrio for final design approval.

andreadelrio reviewed (approved):
Thanks for making the change to use a flyout, @mgiota. Approving now to unblock, but note that we should continue refining this flyout in the future to better align with similar forms-in-flyouts that we already have in Dashboard. For example, take a look at this `Control settings` flyout and the way the button group is styled and accompanied by a label. cc @maciejforcone Please feel free to ping me if you have any questions.

<img width="785" alt="image" src="https://github.com/elastic/kibana/assets/4016496/19dc82df-aead-4105-b313-a5476c193a0d">

mgiota commented:
@elasticmachine merge upstream

mgiota commented:
@elasticmachine merge upstream

simianhacker reviewed (approved):
LGTM

mgiota commented:
@elasticmachine merge upstream

cee-chen reviewed (approved):
🚢
